site/src/pages/WorkspacePage/AppStatuses.tsx

```diff
@@ -0,0 +1,406 @@
+		// Slightly shorter truncation
+		if (fullUrl.length > 40) {
+			const start = fullUrl.slice(0, 20);
+			const end = fullUrl.slice(-20);
+			return `${start}...${end}`;
+		}
+		return fullUrl;
+	} catch {
+		// Slightly shorter truncation
+		if (uri.length > 35) {
+			const start = uri.slice(0, 15);
+			const end = uri.slice(-15);
+			return `${start}...${end}`;
+		}
+		return uri;
+	}
+};
+
+// --- Component Implementation ---
+
+export interface AppStatusesProps {
+	apps: WorkspaceApp[];
+	workspace: Workspace;
+	agents: ReadonlyArray<WorkspaceAgent>;
+	/** Optional reference date for calculating relative time. Defaults to Date.now(). Useful for Storybook. */
+	referenceDate?: Date;
+}
+
+// Extend the API status type to include the app icon and the app itself
+interface StatusWithAppInfo extends APIWorkspaceAppStatus {
+	appIcon?: string; // Kept for potential future use, but we'll primarily use app.icon
+	app?: WorkspaceApp; // Store the full app object
+}
+
+export const AppStatuses: FC<AppStatusesProps> = ({
+	apps,
+	workspace,
+	agents,
+	referenceDate,
+}) => {
+	const theme = useTheme();
+	const { proxy } = useProxy();
+	const preferredPathBase = proxy.preferredPathAppURL;
+	const appsHost = proxy.preferredWildcardHostname;
+
+	// 1. Flatten all statuses and include the parent app object
+	const allStatuses: StatusWithAppInfo[] = apps.flatMap((app) =>
+		app.statuses.map((status) => ({
+			...status,
+			app: app, // Store the parent app object
+		})),
+	);
+
+	// 2. Sort statuses chronologically (newest first)
+	allStatuses.sort(
+		(a, b) =>
+			new Date(b.created_at).getTime() - new Date(a.created_at).getTime(),
+	);
+
+	// Determine the reference point for time calculation
+	const comparisonDate = referenceDate ?? new Date();
+
+	if (allStatuses.length === 0) {
+		return null;
+	}
+
+	return (
+		<div
+			css={{ display: "flex", flexDirection: "column", gap: 16, padding: 16 }}
+		>
+			{allStatuses.map((status, index) => {
+				const isLatest = index === 0;
+				const isFileURI = status.uri?.startsWith("file://");
+				const statusTime = new Date(status.created_at);
+				// Use formatDistance if referenceDate is provided, otherwise formatDistanceToNow
+				const formattedTimestamp = referenceDate
+					? formatDistance(statusTime, comparisonDate, { addSuffix: true })
+					: formatDistanceToNow(statusTime, { addSuffix: true });
+
+				// Get the associated app for this status
+				const currentApp = status.app;
+				let appHref: string | undefined;
+				const agent = agents.find((agent) => agent.id === status.agent_id);
+
+				if (currentApp && agent) {
+					const appSlug = currentApp.slug || currentApp.display_name;
+					appHref = createAppLinkHref(
+						window.location.protocol,
+						preferredPathBase,
+						appsHost,
+						appSlug,
+						workspace.owner_name,
+						workspace,
+						agent,
+						currentApp,
+					);
+				}
+
+				// Determine if app link should be shown
+				const showAppLink =
```

Comment:
Took me awhile to get the avoiding redundant links aspect of this, a comment would go a long way here I think.